packages/mwp-core/src/renderers/browser-render.jsx

```diff
@@ -16,20 +16,28 @@ type AppProps = {
 	basename: string,
 };
 
-/**
- * Async resolver of the props needed for `BrowserApp`:
- * { routes, store, basename }
+/*
+ * Ensures that all async parts of the app that are needed for initial render
+ * are downloaded before returning
  */
 export function resolveAppProps(
 	routes: Array<PlatformRoute>,
 	reducer: Reducer<MWPState, FluxStandardAction>,
 	middleware: Array<Object> = []
 ): Promise<AppProps> {
 	const basename = window.APP_RUNTIME.basename || '';
-	const resolveRoutes = getRouteResolver(routes, basename);
-	const createStore = getBrowserCreateStore(resolveRoutes, middleware);
+	const findMatches = getFindMatches(routes, basename);
+	const createStore = getBrowserCreateStore(findMatches, middleware);
 	const store = createStore(reducer, getInitialState(window.APP_RUNTIME));
-	return resolveRoutes(window.location).then(() => ({
+
+	// find the currently-matched routes that have components that need to
+	// be resolved, then resolve them - this doesn't actually change the route
+	// object, but loads the app chunks into memory
+	const matchedAsyncComponentGetters = findMatches(window.location)
+		.map(({ getComponent }) => getComponent)
+		.filter(x => x);
+
+	return Promise.all(matchedAsyncComponentGetters).then(() => ({
```

Comment:
To make sure i understand, in above we are loading all the chunks into memory and then once all done, actually "resolve" the route now that we have all the pieces. Correct?

Reply:
mostly correct - the 'route' itself is already resolved, but the `getComponent` Promise-returning functions will still be called when the route is rendered by `AsyncRoute`. Since the component modules themselves have already been downloaded, however, the Promises should return instantly.

We actually might want to adjust this behavior to just start rendering immediately, and let the components pop into place as they download - TTI would improve significantly, but it might be a little jarring from a UX perspective. Worth checking out at some point... similar to a FOUC, but maybe not as bad?

Reply:
Yeah maybe but not sure how much its worth if the page feels odd. Our FOM (flash of mobile) already feels awkward. :) I think a change like that we should definitely launch behind a LDF so we can easily test the experience for some users internally :)

Reply:
I've updated this to avoid the flash - it was bad.